### PR TITLE
[docs] Non-language related breaking changes

### LINKED ISSUE #8007: Document changes outside language
Came out of https://github.com/ethereum/solidity/pull/8001. Please see comment here: https://github.com/ethereum/solidity/pull/8001#issuecomment-565096251.

We should probably also list changes outside of the language:
- yul optimizer activated together with the legacy optimizer
- abiv2 not considered experimental
- C API changes
- storage layout export
- metadata hash options
- new error reporter

### PR DESCRIPTION
**Pull request by erak**
Fixes https://github.com/ethereum/solidity/issues/8007.

### CODE REVIEW

File: docs/060-breaking-changes.rst [hunk 1]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+=================
+
+This section lists changes that are unrelated to the language itself, but that have an effect on the interfaces of
+the compiler. These may change the way how you use the compiler on the command line, how you use its programmable
+interface or how you analyze the output produced by it.
+
+New Error Reporter
+~~~~~~~~~~~~~~~~~~
+
+A new error reporter was introduced, which aims on producing more accessible error messages on the command line.
```
Comment (leonardoalt) [inline docs/060-breaking-changes.rst]:
```suggestion
A new error reporter was introduced, which aims at producing more accessible error messages on the command line.
```

File: docs/060-breaking-changes.rst [hunk 2]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+Metadata Hash Options
+~~~~~~~~~~~~~~~~~~~~~
+
+The compiler now appends the `IPFS <https://ipfs.io/>`_ hash of the metadata file to the end of the bytecode by default
+(for details, see documentation on :doc:`contract metadata <metadata>`). Before 0.6.0, the compiler appended the
+`Swarm <https://ethersphere.github.io/swarm-home/>`_ hash by default, and in order to still support this behaviour,
+the new command line option ``--metadata-hash`` was introduced. It allows you to select the hash to be produced and
+appended, by passing either ``ipfs``, ``swarm`` or ``none`` as value to the ``--metadata-hash`` command line option.
+
+These changes also apply to the :ref:`Standard JSON Interface<compiler-api>` and effect the metadata JSON generated by the compiler.
```
Comment (leonardoalt) [inline docs/060-breaking-changes.rst]:
```suggestion
These changes can also be used via the :ref:`Standard JSON Interface<compiler-api>` and affect the metadata JSON generated by the compiler.
```

File: docs/060-breaking-changes.rst [hunk 3]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+It is enabled by default, but passing ``--old-reporter`` falls back to the the deprecated old error reporter.
+
+Metadata Hash Options
+~~~~~~~~~~~~~~~~~~~~~
+
+The compiler now appends the `IPFS <https://ipfs.io/>`_ hash of the metadata file to the end of the bytecode by default
+(for details, see documentation on :doc:`contract metadata <metadata>`). Before 0.6.0, the compiler appended the
+`Swarm <https://ethersphere.github.io/swarm-home/>`_ hash by default, and in order to still support this behaviour,
+the new command line option ``--metadata-hash`` was introduced. It allows you to select the hash to be produced and
+appended, by passing either ``ipfs``, ``swarm`` or ``none`` as value to the ``--metadata-hash`` command line option.
```
Comment (chriseth) [inline docs/060-breaking-changes.rst]: Please also explain that `none` can be used to completely remove it.

File: docs/060-breaking-changes.rst [hunk 4]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+
+These changes also apply to the :ref:`Standard JSON Interface<compiler-api>` and effect the metadata JSON generated by the compiler.
+
```
Comment (chriseth) [inline docs/060-breaking-changes.rst:128]: Maybe add that the safest way is to use the last two bytes to determine the length of the cbor encoding and then do a proper cbor decoding as exlpained in the metadata section (maybe less words also suffice).

File: docs/060-breaking-changes.rst [hunk 5]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+Yul Optimizer
+~~~~~~~~~~~~~
+
+Together with the legacy bytecode optimizer, the :doc:`Yul <yul>` optimizer is now enabled by default when you call the compiler
+with ``--optimize``. If not in assembly mode, you can disable it by calling the compiler with ``--no-optimize-yul``.
```
Comment (chriseth) [inline docs/060-breaking-changes.rst]: Maybe "assembly mode" confuses people more than it helps?

File: docs/060-breaking-changes.rst [hunk 6]
```diff
@@ -98,6 +98,46 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+The option ``--optimize-yul`` is now invalid in assembly mode, since the Yul optimizer is enabled by default.
+
```
Comment (chriseth) [inline docs/060-breaking-changes.rst:135]: Maybe add that this mostly affects the code of the ABIEncoderV2.

File: docs/060-breaking-changes.rst (hunk 3 above)
Comment (axic) [reply docs/060-breaking-changes.rst]: Really? Too bad 😢

File: docs/060-breaking-changes.rst (hunk 4 above)
Comment (Marenz) [reply docs/060-breaking-changes.rst:128]: The safest way to do what? To check if a hash exists? Or to check if it has the expected encoding?

**Comment by Marenz** [comment]
updated

File: docs/060-breaking-changes.rst [hunk 7]
```diff
@@ -98,6 +98,50 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+~~~~~~~~~~~~~~~~~~~~~
+
+The compiler now appends the `IPFS <https://ipfs.io/>`_ hash of the metadata file to the end of the bytecode by default
+(for details, see documentation on :doc:`contract metadata <metadata>`). Before 0.6.0, the compiler appended the
+`Swarm <https://ethersphere.github.io/swarm-home/>`_ hash by default, and in order to still support this behaviour,
+the new command line option ``--metadata-hash`` was introduced. It allows you to select the hash to be produced and
+appended, by passing either ``ipfs`` or ``swarm`` as value to the ``--metadata-hash`` command line option.
+Passing the value ``none`` completely removes the hash.
+
+These can changes also be used via the :ref:`Standard JSON Interface<compiler-api>` and effect the metadata JSON generated by the compiler.
```
Comment (chriseth) [inline docs/060-breaking-changes.rst]:
```suggestion
These features can also be used via the :ref:`Standard JSON Interface<compiler-api>` and effect the metadata JSON generated by the compiler.
```

File: docs/060-breaking-changes.rst [hunk 8]
```diff
@@ -98,6 +98,50 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+
+The recommended way to read the metadata is to read the last two bytes to determine the length of the CBOR encoding
+and perform a proper decoding on that data block as explained in the metadata section.
```
Comment (chriseth) [inline docs/060-breaking-changes.rst]: Can you link to the section?

File: docs/060-breaking-changes.rst [hunk 9]
```diff
@@ -98,6 +98,50 @@ or at least were more difficult to achieve prior to Solidity 0.6.0.
...
+
+Yul Optimizer
+~~~~~~~~~~~~~
+
+Together with the legacy bytecode optimizer, the :doc:`Yul <yul>` optimizer is now enabled by default when you call the compiler
+with ``--optimize``. It can be disabled by calling the compiler with ``--no-optimize-yul``.
+This mostly affects code that uses the ABIEncoderV2.
```
Comment (chriseth) [inline docs/060-breaking-changes.rst]:
```suggestion
This mostly affects code that uses ABIEncoderV2.
```

File: docs/060-breaking-changes.rst (hunk 7 above)
Comment (Marenz) [reply docs/060-breaking-changes.rst]: damn, I was too hasty

**Comment by Marenz** [comment]
Updated

**Review by chriseth — APPROVED** [review]

**Pull request merged**